Added WASM contract interaction via astar.js (#138)
AstarNetwork/astar-docs: docs/builder-guides/XVM and WASM/interact_with_wasm_contract.md

---
sidebar_position: 3
---

# Interact with WASM contract with astar.js
## TL;DR

> During this guide, we'll be guiding you on how to interact with WASM contract with astar.js

Reviewer: Guide twice in the same sentence looks a bit awkward to me.  "In this guide we'll see how…" or "Here we'll be guiding you…" IMO.

---

## What is a astar.js?
Astar.js library provides application developers the ability to query a node and interact with the astar/shiden/shibuya chains using Javascript/Typescript.

## Packages

> The @polkadot/api-contract packages is to be used alongside the @astar-network/astar-api package. With that in mind, we can install from npm, so we are not going to waste too much time with the bare basics.

Reviewer: @polkadot/api-contract package**s** should probable be a singular.

yarn add @astar-network/astar-api@beta @polkadot/api @polkadot/api-contract

## API connection

> The API provides application developers the ability to query astar node for sending transactions to contract.

Reviewer: Maybe it's me, but "query" does not work with "send transaction". "Query data" vs "order to send a transaction". IMO, just "ability to send transaction" would be enough.

if (result.isOk) {
  // should output 123 as per our initial set (output here is an i32)
  console.log('Success', output.toHuman());
} else {
  console.error('Error', result.asErr);
}
```

Underlying the above `.query.<messageName>` is using the `api.rpc.contracts.call` API on the contracts palette to retrieve the value. For this interface, the format is always of the form `messageName(<account address to use>, <value>, <gasLimit>, <...additional params>)`. An example of querying a balance of a specific account on an erc20 contract will therefore be -

```js
// the address we are going to query
const target = '5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY';
const from = '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY';

// only 1 param needed, the actual address we are querying for (more
// params can follow at the end, separated by , if needed by the message)
const callValue = await contract.query.balanceOf(from, { value: 0, gasLimit: -1 }, target);
```

In this example we have specified a `gasLimit` of `-1`, in a subsequent section we will expand on this. for now, just remember that is indicated to use max available, i.e. we don't explicitly want to specify a value.

> When executing it encodes the message using the selector and the input values to allow execution in the contract environment. This can be executed on any contract message, unlike the examples that will follow below it will only read state, not actually execute and therefore not consume any real value from the account.

Reviewer: when executed?

An alternative for reading would be via the lower-level `.read` method, in this case

```javascript
// Perform the actual read (no params at the end, for the `get` message)
// (We perform the send from an account address, it doesn't get executed)
const callValue = await contract
  .read('get', { value, gasLimit })
  .send(alicePair.address);

// The actual result from RPC as `ContractExecResult`
...
```

In cases where the ABI messages have conflicting names, instead of the `'get'` string the actual message index (or message from the Abi itself) can be passed-through.

## Sending a transaction

> In addition to using the `.query.<messageName>` on a contract, the `.tx.<messageName>` method is provides to send an actual encoded transaction to the contract, allow for execution and have this applied in a block. Expanding on our previous examples, we can now execute and then retrieve the subsequent value -

Reviewer: method is provided

```javascript
// We will use these values for the execution
const value = 0; // only useful on isPayable messages
const gasLimit = 3000n * 1000000n;
const incValue = 1;

// Send the transaction, like elsewhere this is a normal extrinsic
// with the same rules as applied in the API (As with the read example,
// additional params, if required can follow - here only one is needed)
await contract.tx
  .inc({ value, gasLimit }, incValue)
  .signAndSend(alicePair, (result) => {
    if (result.status.isInBlock) {
      console.log('in a block');
    } else if (result.status.isFinalized) {
      console.log('finalized');
    }
  });
```

> If we perform the same `query.get` read on the value now, it would be `124`. For lower-level access, like we have in the `Blueprint` via `.createContract` you can also perform the execution via the `.exec` function, which would yield equivalent results -

Reviewer: "we perform" vs "you can perform". Should be consistent throughout the document.

```javascript
// Send the transaction, like elsewhere this is a normal submittable
// extrinsic with the same rules as applied in the API
await contract
  .exec('inc', { value, gasLimit }, incValue)
  .signAndSend(alicePair, (result) => {
    ...
  });
```

For the above interface we can specify the message as the string name, the index of the actual message as retrieved via the Abi.

## Weight estimation

> To estimate the gasLimit (which in the Substrate context refers to the weight used), we can use the `.query` (read) interfaces with a sufficiently large value to retrieve the actual gas consumed. The API makes this easy - with a `gasLimit` or `-1` passed to the query it will use the maximum gas limit available to transactions and the return value will have the actual gas used.

Reviewer: interface (singular)

Where no events were emitted this value would be `undefined`, however should events be emitted, the array will contain all the decoded values.

## Debugging

> One thing you need to know about debugging is Adding #[ink(payable)] to a #[ink(message)] prevents ink_env::debug_println! messages to be logged in console when executing the smart contract call. Debug messages are only emitted during a dry run (query), not during the actual transaction (tx)(Source).When you're calling the contract, first query it, then perform your transaction if there are no error messages.

Reviewer: ```suggestion
One thing you need to remember about debugging is that `#[ink(payable)]` added to an `#[ink(message)]` prevents `ink_env::debug_println!` messages to be logged in console when executing the smart contract call. Debug messages are only emitted during a dry run (query), not during the actual transaction (tx)(Source). When you're calling the contract, first query it, then perform your transaction if there are no error messages.
```
